client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -291,44 +286,82 @@ public final void addReconnectListener(final ReconnectListener<HonoConnection> l
+                    removeFromCheckConnectedPromises(promiseToAdd); // not strictly necessary but keeps the list from growing if connection attempts keep on failing
+                }));
             }
         } else {
             resultHandler.handle(Future.failedFuture(
                     new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
         }
     }
 
+    private boolean addToCheckConnectedPromises(final CheckConnectedPromise promiseToAdd) {
+        // atomically add to checkConnectedPromises - but only if checkConnectedPromises hasn't been set to null
+        final List<CheckConnectedPromise> newPromises = checkConnectedPromises
+                .accumulateAndGet(Collections.singletonList(promiseToAdd), (existing, toAdd) -> {
+                    // no modification of the existing list done here, keeping the accumulatorFunction function side-effect free as required
+                    if (existing == null) {
+                        return null;
+                    }
+                    final List<CheckConnectedPromise> promises = new ArrayList<>(existing.size() + 1);
+                    promises.addAll(existing);
+                    promises.add(toAdd.get(0));
+                    return promises;
+                });
+        return newPromises != null;
+    }
+
+    private void removeFromCheckConnectedPromises(final CheckConnectedPromise promiseToRemove) {
+        checkConnectedPromises
+                .accumulateAndGet(Collections.singletonList(promiseToRemove), (existing, toRemove) -> {
+                    // no modification of the existing list done here, keeping the accumulatorFunction function side-effect free as required
+                    if (existing == null) {
+                        return null;
+                    }
+                    final List<CheckConnectedPromise> promises = new ArrayList<>(existing);
+                    promises.remove(toRemove.get(0));
+                    return promises;
+                });
+    }
+
     private void removeAndCompleteCheckConnectedPromises(final AsyncResult<HonoConnection> connectionResult) {
-        final List<Promise<Void>> promises = checkConnectedPromises.getAndSet(null);
-        if (promises != null) {
+        final List<CheckConnectedPromise> promises = checkConnectedPromises.getAndSet(null);
+        if (promises != null && !promises.isEmpty()) {
+            log.trace("completing {} accumulated connection checks", promises.size());
             final Context ctx = vertx.getOrCreateContext();
-            if (connectionResult.succeeded()) {
-                promises.forEach(promise -> ctx.runOnContext(v -> promise.complete()));
-            } else {
-                promises.forEach(promise -> ctx.runOnContext(v -> promise.fail(connectionResult.cause())));
-            }
+            promises.forEach(promise -> ctx.runOnContext(v -> promise.tryCompleteAndCancelTimer(connectionResult)));
```

Comment:
how about moving all of this into a separate class?

Reply:
I'll have a look...

Reply:
Yes, I've refactored it accordingly.